types/shares.go

```diff
@@ -184,3 +185,311 @@ func zeroPadIfNecessary(share []byte, width int) []byte {
 	}
 	return share
 }
+
+// DataFromSquare extracts block data from an extended data square.
+func DataFromSquare(eds *rsmt2d.ExtendedDataSquare) (Data, error) {
+	originalWidth := eds.Width() / 2
+
+	// sort block data by namespace
+	// define a slice for the raw share data of each type
+	var (
+		// transactions
+		txsShares [][]byte
+		// intermediate state roots
+		isrShares [][]byte
+		// evidence
+		evdShares [][]byte
+		// messages
+		msgShares [][]byte
+	)
+
+	// iterate over each row index
+	for x := uint(0); x < originalWidth; x++ {
+		// iterate over each col index
+		for y := uint(0); y < originalWidth; y++ {
+			// sort the data of that share types via namespace
+			share := eds.Cell(x, y)
+			nid := share[:NamespaceSize]
+			switch {
+			case bytes.Equal(TxNamespaceID, nid):
+				txsShares = append(txsShares, share)
+
+			case bytes.Equal(IntermediateStateRootsNamespaceID, nid):
+				isrShares = append(isrShares, share)
+
+			case bytes.Equal(EvidenceNamespaceID, nid):
+				evdShares = append(evdShares, share)
+
+			case bytes.Equal(TailPaddingNamespaceID, nid):
+				continue
+
+			// every other namespaceID should be a message
+			default:
+				msgShares = append(msgShares, share)
+			}
```

Comment:
This logic doesn't seem to conform to specs. Potential namespace IDs: https://github.com/lazyledger/lazyledger-specs/blob/de5f4f74f56922e9fa735ef79d9e6e6492a2bad1/specs/consensus.md#reserved-namespace-ids. Namespaces up to `NAMESPACE_ID_MAX_RESERVED = 255` are reserved, not messages: https://github.com/lazyledger/lazyledger-specs/blob/de5f4f74f56922e9fa735ef79d9e6e6492a2bad1/specs/consensus.md#constants.

There should also be a check _somewhere_ that a reserved namespace ID that's up to 255 but not specifically reserved is invalid.

Reply:
Thanks for pointing this out! added issue {#67](https://github.com/lazyledger/lazyledger-app/issues/67) to track in the lazyledger-app to add a check in `MsgWirePayForMessage`'s `ValidateBasic` method which get's called with every call to `CheckTx`. 

Added a `case` for this along with a namespace var in `consts.go` to represent the max namespaceID https://github.com/lazyledger/lazyledger-core/blob/d118716769036925491d0412772693a55546ca2c/types/shares.go#L227

Should I also add the const `NamespaceIDMaxReserved` and use that to construct `MaxReservedNamespaceID`?

Reply:
> Should I also add the const `NamespaceIDMaxReserved` and use that to construct `MaxReservedNamespaceID`?

Nah, the setup you have now should be sufficient I think.

Reply:
One additional check that needs to be _somewhere_ is that padding shares between a transaction and a message must have namespace ID `TAIL_TRANSACTION_PADDING_NAMESPACE_ID` (all the way at the bottom of https://github.com/lazyledger/lazyledger-specs/blob/de5f4f74f56922e9fa735ef79d9e6e6492a2bad1/specs/data_structures.md#arranging-available-data-into-shares, under the giant figures)

> The non-interactive default rules may introduce empty shares that do not belong to any message (in the example above, the top-right share is empty). These are zeroes with namespace ID equal to the either `TAIL_TRANSACTION_PADDING_NAMESPACE_ID` if between a request with a reserved namespace ID and a message, or the namespace ID of the previous message if succeeded by a message. See the rationale doc for more info.